keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2861 @@
+means "inject the correct .spec.match during admission and keep it up to date
+with a controller". (Suggested by deads2k). This minimizes version skew if new
+match criteria is added to `spec.match` and also minimizes development effort by
+removing the need to manually declare the fields in CRDs.
+
+The main challenge with this alternative is dealing with mismatches between how
+CRDs declare the `spec.match` schema and what the apiserver expects, even with a
+controller keeping it in sync, it can briefly become out of sync. For this our
+plan is:
+
+- When consuming configuration CRDs OR policy configuration resources used for
+  policy configuration:
+  - If any unrecognized fields, missing required fields, or incorrectly typed
+    fields are found under `spec.match`:
+    - For configuration CRDs:
+      - Set the `ValidatingAdmissionPolicy` state to "misconfigured" in the
+        status (via a Condition, I believe).
+      - Trigger the `FailurePolicy` on all admission validations.
+      - Add a detailed error in the status of the `ValidatingAdmissionPolicy`.
+    - For policy configuration resources:
+      - Track in the status of `ValidatingAdmissionPolicy` that some policy
+        configurations are misconfigured. (also via a Condition?).
+      - Add a detailed error in the status of the `ValidatingAdmissionPolicy`.
+      - Trigger the `FailurePolicy` on admission for resources that match the
+        policy configuration.
+  - If the CRD is deleted:
+    - Set state to "misconfigured
+    - Trigger `FailurePolicy` on all admission validations
+    - Add a detailed error in the status of the `ValidatingAdmissionPolicy`.
+
+A partial `spec.match` schema (subset of the full schema) is okay so long as
+only optional fields are omitted. But any unrecognized field in the `spec.match`
+would not be allowed. 
+
+Proposed annotation: 
+
+> Example: `admission.kubernetes.io/is-policy-configuration-definition: "true"`
+> 
+> Used on: CustomResourceDefinition
+> 
+> What a CustomResourceDefinition has the annotation set to "true", the
+> OpenAPIv3 schema of all versions of this resource is modified and then
+> kept-in-sync by a controller to always contain the expected schema fields of
+> admission policy configuration resources.
+
+xref: https://kubernetes.io/docs/reference/labels-annotations-taints/
+
+Pros:
+
+- A single resource is used to configure both the match criteria and
+  configuration params of a policy.
+- Provides a clear marker on every CRD that is used for policy configuration
+  that it is intended for this purpose, in a machine readable way.
+
+Cons:
+
+- API server must check for a wide range of error conditions and define how
+  exactly it handles each of them.
+- If the `spec.match` schema is incorrectly defined, CRD author might not
+  realize it since they need to check the status of the corresponding
+  `ValidatingAdmissionPolicy` for any errors.
```

Comment:
add: changing this schema in the future could be extremely difficult

Reply:
Added. I've included a few details about the schema fields being atomic.